plugins/node/opentelemetry-koa-instrumentation/src/koa.ts

```diff
@@ -0,0 +1,141 @@
+ * limitations under the License.
+ */
+
+import { BasePlugin } from '@opentelemetry/core';
+import type * as koa from 'koa';
+import * as shimmer from 'shimmer';
+import {
+  KoaMiddleware,
+  KoaContext,
+  KoaComponentName,
+  kLayerPatched,
+} from './types';
+import { VERSION } from './version';
+import { getMiddlewareMetadata } from './utils';
+
+/** Koa instrumentation for OpenTelemetry */
+export class KoaInstrumentation extends BasePlugin<typeof koa> {
+  static readonly component = KoaComponentName;
+  readonly supportedVersions = ['^2.0.0'];
+
+  constructor(readonly moduleName: string) {
+    super('@opentelemetry/koa-instrumentation', VERSION);
+  }
+
+  /**
+   * Patches Koa operations by wrapping the Koa.use function
+   */
+  protected patch(): typeof koa {
+    this._logger.debug('Patching Koa');
+    if (this._moduleExports === undefined || this._moduleExports === null) {
+      return this._moduleExports;
+    }
+    this._logger.debug('Patching Koa.use');
+    shimmer.wrap(this._moduleExports.prototype, 'use', this._getKoaUsePatch);
+
+    return this._moduleExports;
+  }
+
+  /**
+   * Unpatches all Koa operations
+   */
+  protected unpatch(): void {
+    this._logger.debug('Unpatching Koa');
+    shimmer.unwrap(this._moduleExports.prototype, 'use');
+  }
+
+  /**
+   * Patches the Koa.use function in order to instrument each original
+   * middleware layer which is introduced
+   * @param {KoaMiddleware} middleware - the original middleware function
+   */
+  private _getKoaUsePatch(original: (middleware: KoaMiddleware) => koa) {
+    return function use(this: koa, middlewareFunction: KoaMiddleware) {
+      let patchedFunction: KoaMiddleware;
+      if (middlewareFunction.router) {
+        patchedFunction = plugin._patchRouterDispatch(middlewareFunction);
+      } else {
+        patchedFunction = plugin._patchLayer(middlewareFunction, false);
+      }
+      return original.apply(this, [patchedFunction]);
+    };
+  }
+
+  /**
+   * Patches the dispatch function used by @koa/router. This function
+   * goes through each routed middleware and adds instrumentation via a call
+   * to the @function _patchLayer function.
+   * @param {KoaMiddleware} dispatchLayer - the original dispatch function which dispatches
+   * routed middleware
+   */
+  private _patchRouterDispatch(dispatchLayer: KoaMiddleware): KoaMiddleware {
+    this._logger.debug('Patching @koa/router dispatch');
+
+    const router = dispatchLayer.router;
+
+    const routesStack = router?.stack ?? [];
+    for (const pathLayer of routesStack) {
+      const path = pathLayer.path;
+      const pathStack = pathLayer.stack;
+      for (let j = 0; j < pathStack.length; j++) {
+        const routedMiddleware: KoaMiddleware = pathStack[j];
+        pathStack[j] = this._patchLayer(routedMiddleware, true, path);
+      }
+    }
+    const dispatcher: KoaMiddleware = (context: KoaContext, next: koa.Next) => {
+      return dispatchLayer(context, next);
```

Comment:
Can you just call the `displayLayer` directly ? Not sure i get why `dispatcher` exists here

Reply:
Good call - I think this was left over from previous iterations